File: v4-python/phpsaga/step.py
from dataclasses import dataclass, field
from typing import Self
from .failure.strategies import RollbackStrategy, AnyCallable

@dataclass(frozen=True)
class Step:
    name: str
    action: AnyCallable
    compensation: AnyCallable | None = field(default=None, repr=False)
    strategy: RollbackStrategy = RollbackStrategy.LOG_AND_CONTINUE

    @classmethod
    def make(cls, action: AnyCallable, name: str = '') -> 'Step':
        return cls(name=name or getattr(action, '__name__', 'unnamed_step'), action=action)

    def rollback(self, fn: AnyCallable) -> Self:
        # Returns new instance (immutable pattern)
        from dataclasses import replace
        return replace(self, compensation=fn)

    def on_rollback_failure(self, strategy: RollbackStrategy) -> Self:
        from dataclasses import replace
        return replace(self, strategy=strategy)

File: v1-php-core/src/Step.php
<?php declare(strict_types=1);

namespace PhpSaga;

final class Step {
    // Rollback failure strategy constants
    public const RETRY = 1;
    public const LOG_AND_CONTINUE = 2;
    public const ALERT_HUMAN = 3;
    public const THROW = 4;

    private function __construct(
        private readonly string $name,
        private readonly mixed $action,
        private mixed $compensation = null,
        private int $rollbackStrategy = self::LOG_AND_CONTINUE,
    ) {}

    public static function make(callable $action, string $name = ''): self {
        $generatedName = $name;
        if (empty($generatedName)) {
            if (is_string($action)) {
                $generatedName = $action;
            } elseif (is_array($action) && is_string($action[1])) {
                $generatedName = $action[1];
            } else {
                $generatedName = 'unnamed_step';
            }
        }
        return new self($generatedName, $action);
    }

    public function rollback(callable $compensation): self {
        $clone = clone $this;
        $clone->compensation = $compensation;
        return $clone;
    }

    public function onRollbackFailure(int $strategy): self {
        $clone = clone $this;
        $clone->rollbackStrategy = $strategy;
        return $clone;
    }

    /**
     * @throws \Exception
     */
    public function validate(): void {
        if ($this->compensation === null) {
            throw new \Exception("Step '{$this->name}' produces side effects. You MUST define ->rollback()");
        }
    }

    public function getName(): string {
        return $this->name;
    }

    public function getAction(): callable {
        return $this->action;
    }

    public function getCompensation(): ?callable {
        return $this->compensation;
    }

    public function getRollbackStrategy(): int {
        return $this->rollbackStrategy;
    }
}
